### PR TITLE
Develop email registration

### PR DESCRIPTION
**Pull request by rapirent**
- [Add] Added the email registration validation function
- [Add] Added the email reset password function(need email validation first)

the email link will expire in 7 days

### CODE REVIEW

File: smart_campus/app/views.py [hunk 1]
```diff
@@ -18,6 +18,11 @@
 from django.db import IntegrityError
 from django.core import serializers
 from django.core.exceptions import ValidationError
+from django.core.mail import EmailMessage
+from django.template.loader import render_to_string
+from django.utils.http import urlsafe_base64_encode, urlsafe_base64_decode
+from django.utils.encoding import force_bytes, force_text
+from django.contrib.auth.tokens import default_token_generator
```
Comment (Lee-W) [inline smart_campus/app/views.py]: Remove the redundant blank line.

File: smart_campus/app/urls.py [hunk 2]
```diff
@@ -26,5 +26,10 @@
     url('^get_unanswered_question/$', views.get_unanswered_question,
         name='Get Unanswered questsion'),
     url('^add_answered_question/$', views.add_answered_question,
-        name='Add Answered Question')
+        name='Add Answered Question'),
+    url('^activate/(?P<uidb64>[0-9A-Za-z_\-]+)/(?P<token>[0-9A-Za-z]{1,13}-[0-9A-Za-z]{1,20})/$',
+        views.activate, name='activate'),
+    url('^reset_password/(?P<uidb64>[0-9A-Za-z_\-]+)/(?P<token>[0-9A-Za-z]{1,13}-[0-9A-Za-z]{1,20})/$',
+        views.reset_password_page, name='reset password page'),
+    url('^reset_password/$', views.reset_password, name="reset password")
```
Comment (Lee-W) [inline smart_campus/app/urls.py]: Follow the style of views above.
This name should be 'Reset Password'.
Line 31 and 33 should be modified as well.

File: smart_campus/app/views.py [hunk 3]
```diff
@@ -56,6 +62,23 @@ def wrapper(request, *args, **kwargs):
...
+    @wraps(function)
+    def wrapper(request, *args, **kargs):
+        try:
+            user_email = request.POST.get('email')
+            request.user = User.objects.get(email=user_email)
+        except (TypeError, ValueError, OverflowError, User.DoesNotExist):
+            # AnonymousUser
+            pass
+        if request.user.is_anonymous:
+            return HttpResponse('User cannot be found.', status=400)
```
Comment (Lee-W) [inline smart_campus/app/views.py]: 'The user does not exist.'

File: smart_campus/app/views.py [hunk 4]
```diff
@@ -75,10 +98,22 @@ def signup(request):
...
 
+    message = render_to_string('email/activation.html', {
+        'prefix': 'https://' if request.is_secure() else 'http://',
+        'user': user,
+        'domain': request.get_host(),
+        'uid': urlsafe_base64_encode(force_bytes(user.pk)),
+        'token': account_activation_token.make_token(user),
+        'expired_days': settings.PASSWORD_RESET_TIMEOUT_DAYS
+    })
+    mail_subject = '[Activate]Smart Campus App Account Activation'
```
Comment (Lee-W) [inline smart_campus/app/views.py]: '[NCKU Smart Campus App] Please activate your account'

File: smart_campus/app/views.py [hunk 5]
```diff
@@ -1432,3 +1468,60 @@ def add_answered_question(request):
...
+def reset_password(request):
+    message = render_to_string('email/reset_password.html', {
+        'prefix': 'https://' if request.is_secure() else 'http://',
+        'user': request.user,
+        'domain': request.get_host(),
+        'uid': urlsafe_base64_encode(force_bytes(request.user.pk)),
+        'token': default_token_generator.make_token(request.user),
+        'expired_days': settings.PASSWORD_RESET_TIMEOUT_DAYS
+    })
+    mail_subject = '[ResetPassword]Smart Campus App Account Reset Password'
```
Comment (Lee-W) [inline smart_campus/app/views.py]: '[NCKU Smart Campus App] Reset account password'

File: smart_campus/app/templates/app/reset_password_page.html [hunk 6]
```diff
@@ -0,0 +1,50 @@
+{% extends "app/base.html" %}
+{% load static %}
+
+{% block title %}Smart Campus Reset Password {% endblock %}
```
[the remark this answers is not in the document]
Comment (rapirent) [reply smart_campus/app/templates/app/reset_password_page.html]: Removed.

File: smart_campus/app/templates/app/reset_password_page.html [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
```
[the remark this answers is not in the document]
Comment (rapirent) [reply smart_campus/app/templates/app/reset_password_page.html]: Removed.

File: smart_campus/app/urls.py (hunk 2 above)
Comment (rapirent) [reply smart_campus/app/urls.py]: Okay, fixed.

File: smart_campus/app/views.py (hunk 1 above)
Comment (rapirent) [reply smart_campus/app/views.py]: Removed.

File: smart_campus/app/views.py (hunk 3 above)
Comment (rapirent) [reply smart_campus/app/views.py]: Got it.

File: smart_campus/app/views.py (hunk 4 above)
Comment (rapirent) [reply smart_campus/app/views.py]: Modified.

File: smart_campus/app/views.py (hunk 5 above)
Comment (rapirent) [reply smart_campus/app/views.py]: Modified.

**Pull request merged**